core/src/test/java/hudson/UtilTest.java

```diff
@@ -259,21 +269,114 @@ public void testIsSymlink() throws IOException, InterruptedException {
         try {
             File f = new File(d, "f");
-            OutputStream os = new FileOutputStream(f);
+            // Test: File is deleted
+            mkfiles(f);
+            Util.deleteFile(f);
+            assertFalse("f exists after calling Util.deleteFile", f.exists());
+            // Test: If we cannot delete a file, we throw explaining why
+            mkfiles(f);
+            lockFileForDeletion(f);
             try {
                 Util.deleteFile(f);
                 fail("should not have been deletable");
             } catch (IOException x) {
-                assertEquals(c, x.getClass());
-            } finally {
-                os.close();
+                assertThat(calcExceptionHierarchy(x), hasItem(c));
+                assertThat(x.getMessage(), containsString(f.getPath()));
             }
         } finally {
+            unlockFilesForDeletion();
             Util.deleteRecursive(d);
         }
     }
 
     @Test
+    public void testDeleteRecursive() throws Exception {
+        final File tmpDir = tmp.getRoot();
+        final File dir = new File(tmpDir, "dir");
+        final File d1 = new File(dir, "d1");
+        final File d2 = new File(dir, "d2");
+        final File f1 = new File(dir, "f1");
+        final File d1f1 = new File(d1, "d1f1");
+        final File d2f2 = new File(d2, "d1f2");
+        try {
+            // Test: Files get deleted
+            mkdirs(dir, d1, d2);
+            mkfiles(f1, d1f1, d2f2);
+            Util.deleteRecursive(dir);
+            assertFalse("dir exists", dir.exists());
+            // Test: If we cannot delete a file, we throw
+            mkdirs(dir, d1, d2);
+            mkfiles(f1, d1f1, d2f2);
+            lockFileForDeletion(d1f1);
+            try {
+                Util.deleteRecursive(dir);
+                fail("Expected IOException");
+            } catch (IOException x) {
+                assertTrue("dir exists", dir.exists());
+                assertTrue("d1 exists", d1.exists());
+                assertTrue("d1f1 exists", d1f1.exists());
+                assertThat(x.getMessage(), containsString(dir.getPath()));
+            }
+            unlockFilesForDeletion();
+        } finally {
+            unlockFilesForDeletion();
+            Util.deleteRecursive(tmpDir);
+        }
+    }
+
+    /** Creates multiple directories. */
+    private static void mkdirs(File... dirs) {
+        for( File d : dirs ) {
+            d.mkdir();
+            assertTrue(d.getPath(), d.isDirectory());
+        }
+    }
+
+    /** Creates multiple files, each containing their filename as text content. */
+    private static void mkfiles(File... files) throws IOException {
+        for( File f : files )
+            FileUtils.write(f, f.getName());
+    }
+
+    /** Means of unlocking all the files we have locked, indexed by {@link File}. */
+    private static final Map<File, Callable<Void>> unlockFileCallables = new HashMap<File, Callable<Void>>();
```

Comment:
This seems to rely on the fact that methods are not executed in parallel. Better make this and the helper methods that use it non-`static`.